core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -566,13 +534,78 @@ void finishOnRemoteSuccess(Response response) {
             }
         }
 
+        void retryBecauseUnavailable(ShardId shardId, String message) {
+            retry(new UnavailableShardsException(shardId, message + " Timeout: [" + internalRequest.request().timeout() + "], request: " + internalRequest.request().toString()));
+        }
+    }
+
+    /**
+     * Responsible for performing all operations up to the point we start sending requests to replica shards.
+     * When unavailable shards or any retry-able primary errors are encountered they are sent to {@link ReroutePhase}
+     * <p>
+     * Note that as soon as we start sending request to replicas, state responsibility is transferred to {@link ReplicationPhase}
+     */
+    final class PrimaryPhase extends AbstractRunnable {
+        private final InternalRequest internalRequest;
+        private final TransportChannel channel;
+        private final AtomicBoolean finished = new AtomicBoolean(false);
+        private volatile Releasable indexShardReference;
+        private final ClusterStateObserver observer;
+
+        PrimaryPhase(Request request, TransportChannel channel) {
+            this.observer = new ClusterStateObserver(clusterService, request.timeout(), logger);
+            final String concreteIndex;
+            if (resolveIndex()) {
+                concreteIndex = indexNameExpressionResolver.concreteSingleIndex(observer.observedState(), request);
+            } else {
+                concreteIndex = request.index();
+            }
+            this.internalRequest = new InternalRequest(request, concreteIndex);
+            this.channel = channel;
+        }
+
+        @Override
+        public void onFailure(Throwable e) {
+            finishAsFailed(e);
+        }
+
+        @Override
+        protected void doRun() throws Exception {
+            // TODO: we should invoke shards() only on ReroutePhase and
+            // pass the result along the request to the PrimaryPhase to
+            // decouple the logic further
+            final ShardIterator shardIt = shards(observer.observedState(), internalRequest);
+            final ShardRouting primary = resolvePrimary(shardIt);
+            if (primary == null) {
+                finishBecauseUnavailable(shardIt.shardId(), "No active shards.");
+                return;
+            }
+            if (primary.active() == false) {
+                logger.trace("primary shard [{}] is not yet active, scheduling a retry.", primary.shardId());
+                finishBecauseUnavailable(shardIt.shardId(), "Primary shard is not active or isn't assigned to a known node.");
+                return;
+            }
+            if (observer.observedState().nodes().nodeExists(primary.currentNodeId()) == false) {
+                logger.trace("primary shard [{}] is assigned to anode we do not know the node, scheduling a retry.", primary.shardId(), primary.currentNodeId());
```

Comment:
I don't think we need to check for existence, just if it is us, which is covered by the next line. Makes sense?